pkg/controllers/sandbox/restartsandbox_reconciler.go

```diff
@@ -0,0 +1,114 @@
+	verrors "github.com/vertica/vertica-kubernetes/pkg/errors"
+	"github.com/vertica/vertica-kubernetes/pkg/vdbstatus"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// RestartSandboxReconciler will make sure the sandbox and its subclusters'
+// shutdown states are properly set after sandbox restart.
+type RestartSandboxReconciler struct {
+	SRec      *SandboxConfigMapReconciler
+	Vdb       *v1.VerticaDB
+	Log       logr.Logger
+	PFacts    *vdbcontroller.PodFacts
+	RestartSC bool
+}
+
+func MakeRestartSandboxReconciler(r *SandboxConfigMapReconciler,
+	vdb *v1.VerticaDB, pfacts *vdbcontroller.PodFacts, log logr.Logger) controllers.ReconcileActor {
+	return &RestartSandboxReconciler{
+		SRec:   r,
+		Log:    log.WithName("RestartSandboxReconciler"),
+		Vdb:    vdb,
+		PFacts: pfacts,
+	}
+}
+
+func (r *RestartSandboxReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	if res, err := r.updateSandboxShutdownState(ctx); verrors.IsReconcileAborted(res, err) {
+		return res, err
+	}
+	return ctrl.Result{}, r.updateSubclustersShutdownState(ctx)
+}
+
+// updateSandboxShutdownState updates the subcluster's shutdown state after
+// it's been restarted.
+func (r *RestartSandboxReconciler) updateSandboxShutdownState(ctx context.Context) (ctrl.Result, error) {
+	sb := r.Vdb.GetSandbox(r.PFacts.GetSandboxName())
+	// If spec shutdown is true, there's nothing to do
+	// as the operator is still not allowed to restart
+	// the sandbox
+	if sb == nil || sb.Shutdown {
+		return ctrl.Result{}, nil
+	}
+	sbStatus := r.Vdb.GetSandboxStatus(r.PFacts.GetSandboxName())
+	// If status shutdown is equal to spec shutdown, there is nothing
+	// to do
+	if sbStatus == nil || !sbStatus.Shutdown {
+		r.RestartSC = true
+		return ctrl.Result{}, nil
+	}
+	if err := r.PFacts.Collect(ctx, r.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+	// If there is no up node we requeue to first restart the sandbox
+	if r.PFacts.GetUpNodeAndNotReadOnlyCount() == 0 {
+		r.Log.Info("Restart cluster is needed. Requeueing")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	err := vdbstatus.SetSandboxShutdownState(ctx, r.SRec.GetClient(), r.Vdb, r.PFacts.GetSandboxName(), false)
+	r.RestartSC = true
+	return ctrl.Result{}, err
+}
+
+// updateSubclustersShutdownState updates the sandbox's subclusters shutdown state after
+// the sandbox has been restarted.
+func (r *RestartSandboxReconciler) updateSubclustersShutdownState(ctx context.Context) error {
+	if !r.RestartSC {
+		return nil
+	}
+	_, err := vk8s.UpdateVDBWithRetry(ctx, r.SRec, r.Vdb, r.updateSubclustersShutdownStateCallback)
+	return err
+}
+
+func (r *RestartSandboxReconciler) updateSubclustersShutdownStateCallback() (bool, error) {
+	sb := r.Vdb.GetSandbox(r.PFacts.SandboxName)
+	if sb == nil {
+		return false, errors.New("sandbox not found")
+	}
+	needUpdate := false
+	scMap := r.Vdb.GenSubclusterMap()
+	for i := range sb.Subclusters {
```

Comment:
We shouldn't update the spec. Instead, we should update status.subclusters.

Reply:
Either we update spec, or we update both. We must not update status before spec. status update is driven by spec. You change spec, that pushes the operator to do something and then update the status to say that thing was successfully done.
In this case you change `spec.Subcluster[].shutdown` to false, that pushes the operator to look if that subcluster is still down(and restart it if so) , and update `status.Subcluster[].shutdown` as a proof the subcluster is now up.